protobuf/lib/src/protobuf/pb_map.dart

```diff
@@ -31,26 +32,37 @@ class PbMap<K, V> extends MapBase<K, V> {
       : keyFieldType = other.keyFieldType,
         valueFieldType = other.valueFieldType,
         _wrappedMap = Map.unmodifiable(other._wrappedMap),
-        _entryFieldSet = other._entryFieldSet;
+        _entryFieldSet = other._entryFieldSet,
+        _isReadonly = other._isReadonly;
 
   @override
   V operator [](Object key) => _wrappedMap[key];
 
   @override
   void operator []=(K key, V value) {
+    if (_isReadonly)
+      throw new UnsupportedError('Attempted to change a read-only map field');
     _checkNotNull(key);
     _checkNotNull(value);
     _wrappedMap[key] = value;
   }
 
   @override
-  void clear() => _wrappedMap.clear();
+  void clear() {
+    if (_isReadonly)
+      throw new UnsupportedError('Attempted to change a read-only map field');
+    _wrappedMap.clear();
+  }
 
   @override
   Iterable<K> get keys => _wrappedMap.keys;
 
   @override
-  V remove(Object key) => _wrappedMap.remove(key);
+  V remove(Object key) {
+    if (_isReadonly)
+      throw new UnsupportedError('Attempted to change a read-only map field');
```

Comment:
Ideally this would hook into the `frozenMessageModificationHandler`

Reply:
Leaving this out, per discussion.